File: shared/src/jvmTest/kotlin/net/brightroom/uniso/data/repository/AccountRepositoryTest.kt
package net.brightroom.uniso.data.repository

import net.brightroom.uniso.data.model.Account
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class AccountRepositoryTest {
    private lateinit var repository: AccountRepository

    @BeforeTest
    fun setup() {
        val database = createTestDatabase()
        repository = AccountRepository(database)
    }

    @Test
    fun insertAndGetAll() {
        val account = createAccount("acc-1", "x")
        repository.insert(account)

        val accounts = repository.getAll()
        assertEquals(1, accounts.size)
        assertEquals("acc-1", accounts[0].accountId)
        assertEquals("x", accounts[0].serviceId)
    }

    @Test
    fun deleteRemovesAccount() {
        repository.insert(createAccount("acc-1", "x"))
        repository.insert(createAccount("acc-2", "instagram"))
        assertEquals(2, repository.getCount())

        repository.delete("acc-1")
        assertEquals(1, repository.getCount())
        assertEquals("acc-2", repository.getAll()[0].accountId)
    }

    @Test
    fun getAllReturnsSortedBySortOrderThenCreatedAt() {
        repository.insert(createAccount("acc-1", "x", sortOrder = 2, createdAt = "2026-01-01"))
        repository.insert(createAccount("acc-2", "x", sortOrder = 1, createdAt = "2026-01-02"))
        repository.insert(createAccount("acc-3", "x", sortOrder = 1, createdAt = "2026-01-01"))

        val accounts = repository.getAll()
        assertEquals("acc-3", accounts[0].accountId)
        assertEquals("acc-2", accounts[1].accountId)
        assertEquals("acc-1", accounts[2].accountId)
    }

    @Test
    fun getByServiceIdFiltersCorrectly() {
        repository.insert(createAccount("acc-1", "x"))
        repository.insert(createAccount("acc-2", "instagram"))
        repository.insert(createAccount("acc-3", "x"))

        val xAccounts = repository.getByServiceId("x")
        assertEquals(2, xAccounts.size)
        assertTrue(xAccounts.all { it.serviceId == "x" })

        val instaAccounts = repository.getByServiceId("instagram")
        assertEquals(1, instaAccounts.size)
    }

    @Test
    fun getCountReturnsCorrectCount() {
        assertEquals(0, repository.getCount())
        repository.insert(createAccount("acc-1", "x"))
        assertEquals(1, repository.getCount())
        repository.insert(createAccount("acc-2", "instagram"))
        assertEquals(2, repository.getCount())
    }

    @Test
    fun updateDisplayName() {
        repository.insert(createAccount("acc-1", "x", displayName = "Old Name"))
        repository.updateDisplayName("acc-1", "New Name")

        val account = repository.getAll()[0]
        assertEquals("New Name", account.displayName)
    }

    private fun createAccount(
        id: String,
        serviceId: String,
        displayName: String? = null,
        sortOrder: Int = 0,
        createdAt: String = "2026-01-01T00:00:00Z",
    ) = Account(
        accountId = id,
        serviceId = serviceId,
        displayName = displayName,
        avatarUrl = null,
        sortOrder = sortOrder,
        createdAt = createdAt,
    )
}

File: packages/shared/src/__tests__/account-repository.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import type { Database as SqlJsDatabase } from 'sql.js'
import { AccountRepository } from '../data/account-repository'
import { createTestDatabase } from './test-database'

describe('AccountRepository', () => {
  let db: SqlJsDatabase
  let repo: AccountRepository

  beforeEach(async () => {
    db = await createTestDatabase()
    repo = new AccountRepository(db)
  })

  afterEach(() => {
    db.close()
  })

  it('returns empty list when no accounts exist', () => {
    expect(repo.getAll()).toEqual([])
    expect(repo.getCount()).toBe(0)
  })

  it('inserts and retrieves an account', () => {
    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: 'My X',
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    const account = repo.getById('acc-1')
    expect(account).toBeDefined()
    expect(account!.accountId).toBe('acc-1')
    expect(account!.serviceId).toBe('x')
    expect(account!.displayName).toBe('My X')
    expect(repo.getCount()).toBe(1)
  })

  it('getAll returns accounts with service info, ordered by service then sort_order', () => {
    repo.insert({
      accountId: 'acc-ig',
      serviceId: 'instagram',
      displayName: 'IG',
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })
    repo.insert({
      accountId: 'acc-x',
      serviceId: 'x',
      displayName: 'X1',
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    const all = repo.getAll()
    expect(all).toHaveLength(2)
    // X (sort_order 0) comes before Instagram (sort_order 1)
    expect(all[0].serviceId).toBe('x')
    expect(all[1].serviceId).toBe('instagram')
    expect(all[0].serviceDisplayName).toBe('X (旧Twitter)')
    expect(all[0].brandColor).toBe('#000000')
  })

  it('getByServiceId filters correctly', () => {
    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })
    repo.insert({
      accountId: 'acc-2',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 1,
      createdAt: '2026-01-01T00:00:00Z',
    })
    repo.insert({
      accountId: 'acc-3',
      serviceId: 'instagram',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    const xAccounts = repo.getByServiceId('x')
    expect(xAccounts).toHaveLength(2)
    expect(repo.getCountByServiceId('x')).toBe(2)
    expect(repo.getCountByServiceId('instagram')).toBe(1)
  })

  it('getNextSortOrder returns next sequential order', () => {
    expect(repo.getNextSortOrder('x')).toBe(0)

    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    expect(repo.getNextSortOrder('x')).toBe(1)
  })

  it('updates profile', () => {
    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    repo.updateProfile('acc-1', 'Updated Name', 'https://example.com/avatar.png')
    const account = repo.getById('acc-1')
    expect(account!.displayName).toBe('Updated Name')
    expect(account!.avatarUrl).toBe('https://example.com/avatar.png')
  })

  it('updates sort order', () => {
    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    repo.updateSortOrder('acc-1', 5)
    expect(repo.getById('acc-1')!.sortOrder).toBe(5)
  })

  it('deletes an account', () => {
    repo.insert({
      accountId: 'acc-1',
      serviceId: 'x',
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })

    repo.delete('acc-1')
    expect(repo.getById('acc-1')).toBeUndefined()
    expect(repo.getCount()).toBe(0)
  })

  it('getById returns undefined for non-existent account', () => {
    expect(repo.getById('non-existent')).toBeUndefined()
  })
})
